move requires grad into central TLS

robieta opened the pull request:
Coalesce grad mode enabled into the new central TLS. This allows us to skip a redundant TLS lookup in the Autograd kernels. (Since we will need the TLS for AutoDispatchBelowADInplaceOrView as well.) We also inline that function, since it is tiny and on the hot path.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59393
* **#59392**
* #59391
* #59389

Commit:
move requires grad into central TLS

[ghstack-poisoned]

albanD commented on c10/core/GradMode.cpp:
```diff
@@ -3,15 +3,11 @@
-
 bool GradMode::is_enabled() {
-  return GradMode_enabled;
+  return is_enabled(impl::_get_thread_local_state());
```
I guess you don't need this c++ file anymore?

robieta replied:
We do. Several other overloads are defined in it.

albanD reviewed (commented):
cc @ailzhang

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a TLS lookup in the Autograd kernels.



[ghstack-poisoned]

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a TLS lookup in the Autograd kernels.



[ghstack-poisoned]

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a TLS lookup in the Autograd kernels.



[ghstack-poisoned]

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a TLS lookup in the Autograd kernels.



[ghstack-poisoned]

ailzhang commented on c10/core/GradMode.h:
```diff
@@ -1,10 +1,16 @@
 
 struct TORCH_API GradMode {
+  // NB: We inline this overload because it allows us to fully inline hot paths.
+  static inline bool is_enabled(const impl::PODLocalState* const tls) {
```
Would you mind explaining why we want to pass tls as argument instead of just querying `_get_thread_local_state` inside the function body?

robieta replied:
Done.

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a TLS lookup in the Autograd kernels.



[ghstack-poisoned]

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a TLS lookup in the Autograd kernels.



[ghstack-poisoned]

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a redundant TLS lookup in the Autograd kernels. (Since we will need the TLS for AutoDispatchBelowADInplaceOrView as well.) We also inline that function, since it is tiny and on the hot path.



[ghstack-poisoned]

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a redundant TLS lookup in the Autograd kernels. (Since we will need the TLS for AutoDispatchBelowADInplaceOrView as well.) We also inline that function, since it is tiny and on the hot path.



[ghstack-poisoned]

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a redundant TLS lookup in the Autograd kernels. (Since we will need the TLS for AutoDispatchBelowADInplaceOrView as well.) We also inline that function, since it is tiny and on the hot path.



[ghstack-poisoned]

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a redundant TLS lookup in the Autograd kernels. (Since we will need the TLS for AutoDispatchBelowADInplaceOrView as well.) We also inline that function, since it is tiny and on the hot path.



[ghstack-poisoned]

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a redundant TLS lookup in the Autograd kernels. (Since we will need the TLS for AutoDispatchBelowADInplaceOrView as well.) We also inline that function, since it is tiny and on the hot path.



[ghstack-poisoned]

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a redundant TLS lookup in the Autograd kernels. (Since we will need the TLS for AutoDispatchBelowADInplaceOrView as well.) We also inline that function, since it is tiny and on the hot path.



[ghstack-poisoned]

robieta commented:
With respect to the broader TLS question that we've been wrangling with:
Right now a number of components have bits of thread local state. However some of it can be moved across thread boundaries; in effect, one thread can be told to emulate another thread. This stack was meant to be a simple perf optimization; replace multiple TLS lookups with a single lookup of a struct containing centralized TLS data. However it highlighted the fact that we have to shuffling. So the question is do we want to leave the coalesced struct in `c10/core/impl/ThreadLocalState.h` (or whatever we decide to call it), or just abandon all TLS coalescing until someone decides to do a full workup on it?

ezyang commented:
To repeat some of the stuff I said on Tuesday in our sync: it's not unreasonable to setup infrastructure for a change that affects a lot of code, and then slowly migrate stuff onto that infrastructure. So if you don't have time to do the shuffling right now, there might be some utility in setting up the ground rules, and just doing the bare minimum shuffling you need to get performance wins.

That being said, one thing that I'd like to point out about performance optimization is that it is something of an all or nothing deal: identifying a possible optimization is cool, but the software doesn't actually go faster unless you actually land it. So if your goal is to actually make PyTorch go faster, you have to grab the bull by both horns: (1) you have to make the software go faster, and (2) you have to do it in a way that other maintainers are OK with how it's been done. Maybe @swolchok has some other sage advice here.

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a redundant TLS lookup in the Autograd kernels. (Since we will need the TLS for AutoDispatchBelowADInplaceOrView as well.) We also inline that function, since it is tiny and on the hot path.



[ghstack-poisoned]

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a redundant TLS lookup in the Autograd kernels. (Since we will need the TLS for AutoDispatchBelowADInplaceOrView as well.) We also inline that function, since it is tiny and on the hot path.



[ghstack-poisoned]

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a redundant TLS lookup in the Autograd kernels. (Since we will need the TLS for AutoDispatchBelowADInplaceOrView as well.) We also inline that function, since it is tiny and on the hot path.



[ghstack-poisoned]

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a redundant TLS lookup in the Autograd kernels. (Since we will need the TLS for AutoDispatchBelowADInplaceOrView as well.) We also inline that function, since it is tiny and on the hot path.



[ghstack-poisoned]

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a redundant TLS lookup in the Autograd kernels. (Since we will need the TLS for AutoDispatchBelowADInplaceOrView as well.) We also inline that function, since it is tiny and on the hot path.



[ghstack-poisoned]

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a redundant TLS lookup in the Autograd kernels. (Since we will need the TLS for AutoDispatchBelowADInplaceOrView as well.) We also inline that function, since it is tiny and on the hot path.



[ghstack-poisoned]

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a redundant TLS lookup in the Autograd kernels. (Since we will need the TLS for AutoDispatchBelowADInplaceOrView as well.) We also inline that function, since it is tiny and on the hot path.



[ghstack-poisoned]

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a redundant TLS lookup in the Autograd kernels. (Since we will need the TLS for AutoDispatchBelowADInplaceOrView as well.) We also inline that function, since it is tiny and on the hot path.



[ghstack-poisoned]

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a redundant TLS lookup in the Autograd kernels. (Since we will need the TLS for AutoDispatchBelowADInplaceOrView as well.) We also inline that function, since it is tiny and on the hot path.



[ghstack-poisoned]

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a redundant TLS lookup in the Autograd kernels. (Since we will need the TLS for AutoDispatchBelowADInplaceOrView as well.) We also inline that function, since it is tiny and on the hot path.



[ghstack-poisoned]

Commit:
Update on "move requires grad into central TLS"

Coalesce grad mode enabled into the new central TLS. This allows us to skip a redundant TLS lookup in the Autograd kernels. (Since we will need the TLS for AutoDispatchBelowADInplaceOrView as well.) We also inline that function, since it is tiny and on the hot path.



[ghstack-poisoned]
